p2p/host/relay/autorelay.go

```diff
@@ -194,16 +192,83 @@ func (ar *AutoRelay) findRelays(ctx context.Context) {
 		}
 	}
 
+	if haveRelays == 0 {
+		// we failed to find any relays and we are not connected to any!
+		// wait a little and try again, the discovery query might have returned only dead peers
+		log.Debug("no relays connected; retrying in 30s")
+		select {
+		case <-time.After(30 * time.Second):
+			goto again
+		case <-ctx.Done():
+			return
+		}
+	}
+
 	if update > 0 || ar.addrs == nil {
 		ar.updateAddrs()
 	}
 }
 
-func (ar *AutoRelay) selectRelays(pis []pstore.PeerInfo) []pstore.PeerInfo {
+func (ar *AutoRelay) selectRelays(ctx context.Context, pis []pstore.PeerInfo, count, maxq int) []pstore.PeerInfo {
 	// TODO better relay selection strategy; this just selects random relays
 	//      but we should probably use ping latency as the selection metric
+
+	if len(pis) == 0 {
+		return pis
+	}
+
+	if len(pis) < count {
+		count = len(pis)
+	}
+
+	if len(pis) < maxq {
+		maxq = len(pis)
+	}
+
+	// only select relays that can be found by routing
+	type queryResult struct {
+		pi  pstore.PeerInfo
+		err error
+	}
+	result := make([]pstore.PeerInfo, 0, count)
+	resultCh := make(chan queryResult, maxq)
+
+	qctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	// shuffle to randomize the order of queries
 	shuffleRelays(pis)
-	return pis
+	for _, pi := range pis[:maxq] {
+		go func(p peer.ID) {
+			pi, err := ar.router.FindPeer(qctx, p)
```

Comment:
Why?

Reply:
I want to parallelize the `FindPeer`s so that we don't do them serially (and fail slow on dead peers).
I also have a commit I reverted (https://github.com/libp2p/go-libp2p/commit/113dcf91d2cecc7120adfb6b59d5e5a013c2cb01) which uses the addresses returned by discovery -- I can reinstate it.

Reply:
One problem with using the addrs returned by discovery is that it penalizes relay whose peers don't happen to be in the peerstore of the responding peer.
I want to do a true random selection over _live_ relays and that short-circuits.

Reply:
Also, hotspot creation.

Reply:
Hrm, that might be ok -- although it does prefer relays for which we have addrs, and this may include failed connection attempts in a retry because `TempAddrTTL` is now long enough to cause problems.
But then again, these will be results form a `FindPeer` query so we might as well retry some relays to see if they were temporarily down (assuming they were up because they were found).

Reply:
done in https://github.com/libp2p/go-libp2p/pull/598/commits/a8d14f9b025d3073c60938257735e60ff8e23b89

Reply:
Perhaps we should _add_ the addrs from the result of the query in the peerstore too, but I don't know what's the right TTL here.

Reply:
Random idea: we could recompute the relay's addresses from our peerstore every time the user calls Addrs. That way, we update them based on new addresses from the relay. That'll make this code significantly less important (unnecessary?).

Reply:
we'll still have to clean up addrsploded relay, but we need this code so that we put a strict upper bound in the discovery time of relay addresses from parallelization.
Otherwise, we'll have to do it serially for every relay we tried (the old code) and be very slow when we get dead relays in the query result.

Reply:
I think I'll bite the bullet and use addrs from discovery; we are shuffling anyway.
This should reduce the number of `FindPeers` queries with large sets.

Reply:
but I still think it's a bad idea; the quality of addrs we get from discovery is very poor.
flip-flop.

Reply:
ah, it's unnecessary, `FindProviders` stashes in the peerstore anyway. So yeah, nothing we can do anyway.

Reply:
I think we can use a blacklist to avoid using the same peers within the context of a single `findRelays` if we failed to connect.

Reply:
maybe not a good idea, as a relay may be temporarily down.

Reply:
but on the same time, it's just a single pass over the relay space and the known addrs short-circuit might result in the same (dead) relays being hit all over again.

Reply:
I added the ignore list, which gets cleaned up between retries if we fail to select any relays.

Reply:
and I reverted the whole thing, as it exhibits very complex behaviour that is biased towards full DHT clients and penalizes dedicated relays like ours.